File: src/main/java/ch/grassl/day1/impl/ElfImporter.java
package ch.grassl.day1.impl;

import ch.grassl.common.Importer;
import ch.grassl.util.ResourceReader;

import java.util.ArrayList;
import java.util.List;

import static ch.grassl.util.Util.isEmpty;

public class ElfImporter implements Importer<Elf> {

    public static ElfImporter getInstance() {
        return new ElfImporter();
    }
    @Override
    public List<Elf> importData(String resource) {
        String[] data = ResourceReader.of(resource).read();
        return mapToElves(data);
    }

    private static List<Elf> mapToElves(String[] data) {
        List<Elf> elves = new ArrayList<>();
        for (int i = 0; i < data.length; i++) {
            if (!isEmpty(data[i])) {
                int calories = 0;
                for (int j = i; j < data.length; j++) {
                    if (!isEmpty(data[j])) {
                        calories += Integer.parseInt(data[j]);
                    } else {
                        i = j;
                        break;
                    }
                }
                elves.add(new Elf(calories));
            }
        }
        return elves;
    }
}

File: src/main/java/ch/grassl/day04/impl/ElfImporter.kt
package ch.grassl.day04.impl

import ch.grassl.common.Importer
import ch.grassl.util.ResourceReader

class ElfImporter : Importer<Pair<Elf, Elf>> {

    override fun importData(resource: String): List<Pair<Elf, Elf>> {
        val lines = ResourceReader.of(resource).read()
        return mapToPairs(lines)
    }

    private fun mapToPairs(lines: Array<String>): List<Pair<Elf, Elf>> {
        return lines
            .map { line ->
                val section = line.split(",")
                section
                    .map { it.split("-") }
                    .map { IntRange(it[0].toInt(), it[1].toInt()) }}
            .map { Pair(Elf(it[0]), Elf(it[1])) }
    }
}